### PR TITLE
Feature/archcnl 68/simple architecture GUI

### PR DESCRIPTION
**Pull request by cmkehler**
Added a simple GUI to add new ArchitectureRules via an TextArea; to be expanded in the future.
The rules will get added to the ArchitectureRuleManager, but aren't checked for correctness yet.
I will add a presenter to NewArchitectureRuleView.java on request (I am unsure if one is necessary in this case)

### CODE REVIEW

File: web-ui/src/main/java/org/archcnl/ui/input/ruleeditor/ArchitectureRulesLayout.java [hunk 1]
```diff
@@ -8,29 +8,32 @@
...
-                    "Architecture Rules",
-                    "Create new Arch Rule",
-                    new ButtonClickResponder() {
-                        @Override
-                        public void onButtonClick() {}
-                    });
     VerticalLayout rulesLayout = new VerticalLayout();
 
-    public ArchitectureRulesLayout() {
+    public ArchitectureRulesLayout(InputView Parent, MainPresenter mainPresenter) {
```
Comment (VincentRaudszus) [inline web-ui/src/main/java/org/archcnl/ui/input/ruleeditor/ArchitectureRulesLayout.java:23]: The parameter `mainPresenter` is not used here and thus should not be handed over. Additionally, the parameter `parent` should be lowercase according to the style conventions.

File: web-ui/src/main/java/org/archcnl/ui/input/ruleeditor/ArchitectureRulesLayout.java [hunk 2]
```diff
@@ -8,29 +8,32 @@
...
         RulesConceptsAndRelations.getInstance()
                 .getArchitectureRuleManager()
                 .addPropertyChangeListener(this);
         // Remove style property to makes no sense in this layout
         // TODO: Separate ArchitectureRulesLayout from CreateNewLayout
+
+        ButtonClickResponder ruleCreationClickResponder = this::switchToNewArchitectureRuleView;
```
Comment (VincentRaudszus) [inline web-ui/src/main/java/org/archcnl/ui/input/ruleeditor/ArchitectureRulesLayout.java:30]: You can get rid of this variable by directly inserting the `this::switchToNewArchitectureRuleView` into the `CreateNewLayout` constructor.

File: web-ui/src/main/java/org/archcnl/ui/input/ruleeditor/ArchitectureRulesLayout.java [hunk 3]
```diff
@@ -8,29 +8,32 @@
...
-                    });
     VerticalLayout rulesLayout = new VerticalLayout();
 
-    public ArchitectureRulesLayout() {
+    public ArchitectureRulesLayout(InputView Parent, MainPresenter mainPresenter) {
         RulesConceptsAndRelations.getInstance()
                 .getArchitectureRuleManager()
                 .addPropertyChangeListener(this);
         // Remove style property to makes no sense in this layout
         // TODO: Separate ArchitectureRulesLayout from CreateNewLayout
```
Comment (VincentRaudszus) [inline web-ui/src/main/java/org/archcnl/ui/input/ruleeditor/ArchitectureRulesLayout.java:28]: These two comment lines belong to the line `createNewRuleLayout.getStyle().remove("border");` and thus should be left directly above that line.

File: web-ui/src/main/java/org/archcnl/ui/input/ruleeditor/ArchitectureRulesLayout.java [hunk 4]
```diff
@@ -49,6 +52,10 @@ private void updateRules() {
         }
     }
 
+    public void switchToNewArchitectureRuleView() {
```
Comment (VincentRaudszus) [inline web-ui/src/main/java/org/archcnl/ui/input/ruleeditor/ArchitectureRulesLayout.java:55]: I would prefer having the rule creation view integrated into the normal `ArchitectureRulesLayout` so the user can see the already existing rules while creating a new one. But this implementation is fine for now.

File: web-ui/src/main/java/org/archcnl/ui/input/ruleeditor/NewArchitectureRuleView.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+import org.archcnl.ui.input.RulesOrMappingEditorView;
+
+public class NewArchitectureRuleView extends RulesOrMappingEditorView {
+
+    private static final long serialVersionUID = -2966045554441002128L;
+    private Button saveButton;
+    private TextArea archRuleTextArea;
+    private InputView parent;
+
+    public NewArchitectureRuleView(InputView Parent) {
```
Comment (VincentRaudszus) [inline web-ui/src/main/java/org/archcnl/ui/input/ruleeditor/NewArchitectureRuleView.java:19]: Again `Parent` should be lowercase

File: web-ui/src/main/java/org/archcnl/ui/input/ruleeditor/NewArchitectureRuleView.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        parent = Parent;
+        getStyle().set("overflow", "auto");
+        getStyle().set("border", "1px solid black");
+
+        saveButton = new Button("Save Rule", e -> saveRule());
+        archRuleTextArea = new TextArea("", "Every Aggregate must residein a DomainRing");
+        archRuleTextArea.setWidthFull();
+
+        add(new Label("Write new Adoc Rule"));
```
Comment (VincentRaudszus) [inline web-ui/src/main/java/org/archcnl/ui/input/ruleeditor/NewArchitectureRuleView.java:28]: 1. The user is creating an architecture rule or arch rule, not an "Adoc Rule"
2. The label property of the `archRuleTextArea` could be used making this `Label` unnecessary.

File: web-ui/src/main/java/org/archcnl/ui/input/ruleeditor/NewArchitectureRuleView.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+            } catch (NoArchitectureRuleException e) {
+                e.printStackTrace();
+            }
+        }
+        parent.switchToArchitectureRulesView();
+    }
+
+    private ArchitectureRule parseArchitectureRule(String potentialRule)
+            throws NoArchitectureRuleException {
+        if (false) {
```
Comment (VincentRaudszus) [inline web-ui/src/main/java/org/archcnl/ui/input/ruleeditor/NewArchitectureRuleView.java:49]: I think it's fine to leave this if-statement out for now as it only causes compiler warnings ("dead code").

**Review by VincentRaudszus — CHANGES_REQUESTED** [review]
Additionally to the comments, I added in the code I would like to request the addition of a contracted presenter for the `NewArchitectureRuleView`.

File: web-ui/src/main/java/org/archcnl/ui/input/ruleeditor/ArchitectureRulesContract.java [hunk 8]
```diff
@@ -0,0 +1,16 @@
+package org.archcnl.ui.input.ruleeditor;
+
+import java.io.Serializable;
+
+public interface ArchitectureRulesContract {
+
+    public interface View {
+        public void saveRule();
```
Comment (VincentRaudszus) [inline web-ui/src/main/java/org/archcnl/ui/input/ruleeditor/ArchitectureRulesContract.java:8]: The contract does not act as an interface description for the View and the Presenter. Its only purpose is to describe which actions the view is allowed to perform on the presenter and the other way around.
The method `saveRule` is only meant to be called from within the view and never from the presenter. This means that it should not be part of the contract between the View and the Presenter.

Comment (cmkehler) [reply web-ui/src/main/java/org/archcnl/ui/input/ruleeditor/ArchitectureRulesContract.java]: I thought that maybe in the future there is a scenario where the presenter might want to read out the textarea. But thinking about it it seems unlikely. In any case the method should only be part of the contract in the case that the presenter uses it, so I removed it.

**Pull request merged**